scribus/plugins/import/pdf/slaoutput.cpp

```diff
@@ -3896,3 +3927,604 @@ bool SlaOutputDev::checkClip()
 	}
 	return ret;
 }
+
+
+/*
+In geometry, collinearity of a set of points is the property of their lying on a single line. A set of points with this property is said to be collinear.
+In greater generality, the term has been used for aligned objects, that is, things being "in a line" or "in a row".
+ PDF never deviates from the line when it comes to collinear, but allow for 1pixel of divergence
+*/
+bool TextRegion::collinear(qreal a, qreal b)
+{
+	return abs(a - b) < 1 ? true : false;
+}
+
+// like collinear but we allow a deviation of up to +-2 rejoin font text widths
+bool TextRegion::isCloseToX(qreal x1, qreal x2)
+{
+	//FIXME: This should use the char width not linespacing which is y
+	return (abs(x2 - x1) <= lineSpacing * 6) || (abs(x1 - this->textRegioBasenOrigin.x()) <= lineSpacing);
+}
+
+// like collinear but we allow a deviation of up to 3 rejoin font linespaces, but in one direction and half a line space in the other direction
+bool TextRegion::isCloseToY(qreal y1, qreal y2) 
+{
+	//FIXME: Actually test the correct magnitudes not the abs value. There should be a parameter in the ui to set the matching tolerance but hard code to allow 3 linespaces difference before we say that the y is out of scope.
+	return (y2 - y1) >= 0 && y2 - y1 <= lineSpacing * 3;
+}
+
+// less than the last y value but bot more than the line spacing less, could also use the base line of the last line to be more accurate
+bool TextRegion::adjunctLesser(qreal testY, qreal lastY, qreal baseY) 
+{
+	return (testY > lastY
+		&& testY <= baseY + lineSpacing
+		&& lastY <= baseY + lineSpacing);
+}
+
+// less than the last y value but bot more than the line spacing less, could also use the base line of the last line to be more accurate
+bool TextRegion::adjunctGreater(qreal testY, qreal lastY, qreal baseY)
+{
+	return (testY <= lastY
+		&& testY >= baseY - lineSpacing * 0.75
+		&& lastY != baseY);
+}
+
+TextRegion::FrameworkLineTests TextRegion::linearTest(QPointF point, bool xInLimits, bool yInLimits)
+{
+	//TODO: add a FIRSTPOINT result as well
+	FrameworkLineTests pass = FrameworkLineTests::FAIL;

```

Comment:
if i get it correctly, each branch set the value that will be returned, which won't be changed anymore.

instead of setting the value at the beginning, and returning at the end, just return as soon as you have the value.

i have removed all comments and debug to be able to reason with this...

```cpp
	if (collinear(point.y(), lastXY.y()))
		if (collinear(point.x(), lastXY.x())) 
			return FrameworkLineTests::FIRSTPOINT;
		else if (xInLimits) 
			return FrameworkLineTests::SAMELINE;
	else if (adjunctLesser(point.y(), lastXY.y(), lineBaseXY.y())) 
		return FrameworkLineTests::STYLESUPERSCRIPT;
	else if (adjunctGreater(point.y(),lastXY.y(),lineBaseXY.y()))
		if (collinear(point.y(), lineBaseXY.y()))
			return FrameworkLineTests::STYLENORMALRETURN;
		else
			return FrameworkLineTests::STYLESUPERSCRIPT;
	else
		if (isCloseToX(point.x(), textRegioBasenOrigin.x()))
			if (isCloseToY(point.y(), lastXY.y()) && !collinear(point.y(), lastXY.y()))
				if ((textRegionLines.size() >= 2))
					return FrameworkLineTests::NEWLINE;
				else if (textRegionLines.size() == 1) 
					return FrameworkLineTests::NEWLINE;
		else
			qDebug() << "NEWLINE2 oops:"<<point << ":"<<textRegioBasenOrigin << ":" << lineSpacing;
	return FrameworkLineTests::FAIL;
```

the first think i recognize is that the last else with the fail should be a top level else (it works as is... but...)

```cpp
	if (collinear(point.y(), lastXY.y()))
		if (collinear(point.x(), lastXY.x())) 
			return FrameworkLineTests::FIRSTPOINT;
		else if (xInLimits) 
			return FrameworkLineTests::SAMELINE;
	else if (adjunctLesser(point.y(), lastXY.y(), lineBaseXY.y())) 
		return FrameworkLineTests::STYLESUPERSCRIPT;
	else if (adjunctGreater(point.y(),lastXY.y(),lineBaseXY.y()))
		if (collinear(point.y(), lineBaseXY.y()))
			return FrameworkLineTests::STYLENORMALRETURN;
		else
			return FrameworkLineTests::STYLESUPERSCRIPT;
	else if (isCloseToX(point.x(), textRegioBasenOrigin.x())) &&(isCloseToY(point.y(), lastXY.y()) && !collinear(point.y(), lastXY.y()))
		if ((textRegionLines.size() >= 2))
			return FrameworkLineTests::NEWLINE;
		else if (textRegionLines.size() == 1) 
			return FrameworkLineTests::NEWLINE;
	else
		qDebug() << "NEWLINE2 oops:"<<point << ":"<<textRegioBasenOrigin << ":" << lineSpacing;
	return FrameworkLineTests::FAIL;
```

this makes stick out that the `!collinear(point.y(), lastXY.y())` in the last `else if` is already ensure by the first `if`, that the last else is not needed anymore and that it's better to first check `== 1` and then `>=  2`

```cpp
	if (collinear(point.y(), lastXY.y()))
		if (collinear(point.x(), lastXY.x())) 
			return FrameworkLineTests::FIRSTPOINT;
		else if (xInLimits) 
			return FrameworkLineTests::SAMELINE;
	else if (adjunctLesser(point.y(), lastXY.y(), lineBaseXY.y())) 
		return FrameworkLineTests::STYLESUPERSCRIPT;
	else if (adjunctGreater(point.y(),lastXY.y(),lineBaseXY.y()))
		if (collinear(point.y(), lineBaseXY.y()))
			return FrameworkLineTests::STYLENORMALRETURN;
		else
			return FrameworkLineTests::STYLESUPERSCRIPT;
	else if (isCloseToX(point.x(), textRegioBasenOrigin.x()) && isCloseToY(point.y(), lastXY.y()))
		if (textRegionLines.size() == 1) 
			return FrameworkLineTests::NEWLINE;
		else if ((textRegionLines.size() >= 2))
			return FrameworkLineTests::NEWLINE;
	qDebug() << "NEWLINE2 oops:"<<point << ":"<<textRegioBasenOrigin << ":" << lineSpacing;
	return FrameworkLineTests::FAIL;
```

now, personally, i can read most of the lines and i understand them better without the comments... i'll leave it to you to decide if you want to refactor the code this way and how you want to re-add the comments and debugs.  
for the comments and todo, personally, i would prefer to see them before the function definition in a nice big comment. just ask if you don't understand what i mean by this.

p.s: i hope that my edits are correct! please check carefully the code above before using it: i did not run or compile it!

Reply:
done, i've removed all the comments and debug and placed a larger descripive comment in relation to the fuynction